Ability to define custom metadata + some documentation errors

LennDG opened the pull request:
I added 2 .tmpl files to give the possibility to add custom metadata to features. This will override all regular metadata. The PR does not create any breaking changes to already existing code that uses the project.

One bug that I do not know how to fix is that when many variables are added, the list of Metadata variables in a Feature overview will start continuing outside of the Metadata overview box.

I added the documentation on how this feature is used and corrected some typos and errors in the already existing documentation.

wswebcreation commented on lib/generate-report.js:
```diff
@@ -367,11 +371,16 @@ function generateReport(options) {
         const featuresOverviewIndex = path.resolve(reportPath, INDEX_HTML);
 
+        // Assign the correct overview template
+        var featuresOverviewTemplate;
```
Please write this as and try to avoid the usage of `var`

`FEATURE_METADATA_OVERVIEW_TEMPLATE = suite.customMetadata ? FEATURE_CUSTOM_METADATA_OVERVIEW_TEMPLATE : FEATURE_METADATA_OVERVIEW_TEMPLATE;`

wswebcreation commented on lib/generate-report.js:
```diff
@@ -399,6 +408,12 @@ function generateReport(options) {
     function _createFeatureIndexPages(suite) {
+
+        // Set custom metadata overview for the feature
+        var featuresMetadataOverview;
```
Please write this as below and try to avoid the usage of `var`

`FEATURE_METADATA_OVERVIEW_TEMPLATE = suite.customMetadata ? FEATURE_CUSTOM_METADATA_OVERVIEW_TEMPLATE : FEATURE_METADATA_OVERVIEW_TEMPLATE;`

wswebcreation commented on lib/generate-report.js:
```diff
@@ -16,10 +16,12 @@ const FEATURE_FOLDER = 'features';
 const FEATURES_OVERVIEW_INDEX_TEMPLATE = 'features-overview.index.tmpl';
 const CUSTOMDATA_TEMPLATE = 'components/custom-data.tmpl';
 const FEATURES_OVERVIEW_TEMPLATE = 'components/features-overview.tmpl';
```
Make a `let` of this

wswebcreation commented on lib/generate-report.js:
```diff
@@ -16,10 +16,12 @@ const FEATURE_FOLDER = 'features';
 const FEATURES_OVERVIEW_INDEX_TEMPLATE = 'features-overview.index.tmpl';
 const CUSTOMDATA_TEMPLATE = 'components/custom-data.tmpl';
 const FEATURES_OVERVIEW_TEMPLATE = 'components/features-overview.tmpl';
+const FEATURES_OVERVIEW_CUSTOM_METADATA_TEMPLATE = 'components/features-overview-custom-metadata.tmpl'
```
Your missing a `;` here

wswebcreation commented on lib/generate-report.js:
```diff
@@ -16,10 +16,12 @@ const FEATURE_FOLDER = 'features';
 const FEATURES_OVERVIEW_CHART_TEMPLATE = 'components/features-overview.chart.tmpl';
 const SCENARIOS_OVERVIEW_CHART_TEMPLATE = 'components/scenarios-overview.chart.tmpl';
 const FEATURE_OVERVIEW_INDEX_TEMPLATE = 'feature-overview.index.tmpl';
 const FEATURE_METADATA_OVERVIEW_TEMPLATE = 'components/feature-metadata-overview.tmpl';
```
Make a `let` of this

wswebcreation commented on lib/generate-report.js:
```diff
@@ -367,11 +371,16 @@ function generateReport(options) {
             _.template(_readTemplateFile(FEATURES_OVERVIEW_INDEX_TEMPLATE))({
                 suite: suite,
-                featuresOverview: _.template(_readTemplateFile(FEATURES_OVERVIEW_TEMPLATE))({
+                featuresOverview: _.template(_readTemplateFile(featuresOverviewTemplate))({
```
Keep the old line here because we re-assigned the `let`

wswebcreation commented on lib/generate-report.js:
```diff
@@ -409,7 +424,7 @@ function generateReport(options) {
                         _: _
                     }),
-                    featureMetadataOverview: _.template(_readTemplateFile(FEATURE_METADATA_OVERVIEW_TEMPLATE))({
+                    featureMetadataOverview: _.template(_readTemplateFile(featuresMetadataOverview))({
```
Keep the old line here because we re-assigned the `let`

wswebcreation commented on test/test.js:
```diff
@@ -4,9 +4,15 @@ const test = require('../lib/generate-report');
 test.generate({
     saveCollectedJSON: true,
-    jsonDir: './test/unit/data/json/',
+    jsonDir: './reports/',
```
Please provide a new test file with new data and and a new location so we can generate the report for your specific changes

LennDG commented:
Alright I will incorporate your comments. Quite new to Javascript so I will take some time to figure out best practices.

LennDG commented:
Sorry for the late update, I was sick last week.

The first issue is because the customMetadata flag was set to true for the old data. Since the old does not have metadata formed as [{name: "foo", value: "bar"}] it has some weird effects. In correct usage this should not happen. I added a new test and changed the variable assignments like you requested.

wswebcreation commented on test/unit/data/custom-metadata-json/ambiguous_scenarios_specified_v1.json:
```diff
@@ -0,0 +1,100 @@
+[
```
Hi @LennDG 

Tnx for the changes. I only think you missed something. If I now generate the files there is no `custom-metadata` in the report, there is also no custom metadata in the pushed JSON files

LennDG replied:
I overwrite the metadata of the JSON files and have not added a `custom-metadata` field. This is to use the existing code as much as possible without needing changes.

wswebcreation commented on test/custom-metadata-test.js:
```diff
@@ -0,0 +1,21 @@
+'use strict';
```
When I now look at this I think we can do it better like this in 1 file. I've also adjusted the paths so they will not overwrite each other. Hope you don't mind it.

```js
'use strict';

const test = require('../lib/generate-report');

/**
 * Generate a report for browsers
 */
test.generate({
    saveCollectedJSON: true,
    jsonDir: './test/unit/data/json/',
    reportPath: './.tmp/browsers/',
    reportName: 'You can adjust this report name',
    customData: {
        title: 'Run info',
        data: [
            {label: 'Project', value: 'Custom project'},
            {label: 'Release', value: '1.2.3'},
            {label: 'Cycle', value: 'B11221.34321'},
            {label: 'Execution Start Time', value: 'Nov 19th 2017, 02:31 PM EST'},
            {label: 'Execution End Time', value: 'Nov 19th 2017, 02:56 PM EST'}
        ]
    }
});

/**
 * Generate a report with custom metadata
 */
test.generate({
    saveCollectedJSON: true,
    jsonDir: './test/unit/data/custom-metadata-json/',
    reportPath: './.tmp/custom-metadata/',
    reportName: 'You can adjust this report name',
    customMetadata: true,
    customData: {
        title: 'Run info',
        data: [
            {label: 'Project', value: 'Custom project'},
            {label: 'Release', value: '1.2.3'},
            {label: 'Cycle', value: 'B11221.34321'},
            {label: 'Execution Start Time', value: 'Nov 19th 2017, 02:31 PM EST'},
            {label: 'Execution End Time', value: 'Nov 19th 2017, 02:56 PM EST'}
        ]
    }
});
```

LennDG replied:
It's your project, I can't tell you what to do with filepaths :)

wswebcreation commented on test/unit/data/custom-metadata-json/ambiguous_scenarios_specified_v1.json:
```diff
@@ -0,0 +1,100 @@
+    "name": "Ambiguous scenarios specified V1",
+    "tags": [],
+    "uri": "/Users/wswebcreation/deTesters/protractor-cucumber-typescript-boilerplate/e2e-tests/features/ambiguous.feature",
+    "metadata": [
```
I think this can be removed

LennDG replied:
Do you mean that the file can be removed? I just overwrote the `metadata` field to re use it.

wswebcreation replied:
No this is good 👍

wswebcreation commented on test/unit/data/custom-metadata-json/ambiguous_scenarios_specified_v2.json:
```diff
@@ -0,0 +1,72 @@
+[
+  {
+    "metadata": [
```
I think this can be removed

wswebcreation commented on test/unit/data/custom-metadata-json/happy_flow_app_android.json:
```diff
@@ -0,0 +1,77 @@
+[
+  {
+    "metadata": [
```
I think this can be removed, I also think the `App` needs to be removed from the testdata so we have "clean" data

LennDG replied:
I cleaned out the App because it indeed didn't make sense and was confusing.

wswebcreation commented on test/unit/data/custom-metadata-json/pending_scenarios_specified_v1.json:
```diff
@@ -0,0 +1,102 @@
+    "name": "Pending scenarios specified V1",
+    "tags": [],
+    "uri": "/Users/wswebcreation/deTesters/protractor-cucumber-typescript-boilerplate/e2e-tests/features/pending.feature",
+    "metadata": [
```
I think this can be removed

wswebcreation commented on test/unit/data/custom-metadata-json/skipped_scenarios.json:
```diff
@@ -0,0 +1,278 @@
+[
+  {
+    "metadata": [
```
I think this can be removed

wswebcreation commented on test/unit/data/custom-metadata-json/undefined_scenarios_specified_v3.json:
```diff
@@ -0,0 +1,90 @@
+[
+  {
+    "metadata": [
```
I think this can be removed

LennDG commented:
I will update the test data to reflect other use cases than browser versions.

LennDG commented:
The pushed JSON files have their custom metadata under the `metadata` field. I decided against using a `custom-metadata` field to prevent users being confused by trying to use both simultaneously, which does not work.

wswebcreation commented:
Hi @LennDG 

It now looks great. Thanks for this feature! I'll merge it and make a new release
